File: broodling/contract.py
"""The immutable Work Contract revision.

A Contract is a source-attributed statement of what the Work Unit must satisfy.
It is a *record*, not a filter: an obligation that V1 cannot support is
representable here verbatim so that Closability can reject it while preserving
it. Nothing in this module removes or weakens a stated obligation.

Canonical bytes are the durable identity of a revision. Two Contracts with the
same meaning produce the same bytes and therefore the same revision; a
meaning-changing replacement produces different bytes and therefore a new
revision.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .identity import digest

CONTRACT_FORMAT = 1

#: How a Contract's structure was produced. Model extraction is permitted to
#: propose structure; it is recorded, and it never entitles a source.
CONSTRUCTED_BY = frozenset({"broodling_policy", "caller", "model_extraction"})


@dataclass(frozen=True, slots=True)
class SourceAttribution:
    """A pinned entitled source this Contract was built from."""

    source_id: str
    content_sha256: str

    def to_mapping(self) -> dict[str, Any]:
        return {"sourceId": self.source_id, "contentSha256": self.content_sha256}


@dataclass(frozen=True, slots=True)
class EvidencePopulation:
    """Optional validation guidance, preserved without a boundedness admission gate."""

    kind: str
    members: tuple[str, ...] = ()
    surface: str = ""

    def to_mapping(self) -> dict[str, Any]:
        return {
            "kind": self.kind,
            "members": list(self.members),
            "surface": self.surface,
        }


@dataclass(frozen=True, slots=True)
class MechanicalEvidence:
    """A frozen check declaration supplied to Zeroshot as task context.

    Retained for Contract compatibility. Broodling does not execute the argv or
    construct an independent evidence record; the standard native workflow owns
    implementation and verification.
    """

    argv: tuple[str, ...]
    cwd: str = "."
    materials: tuple[str, ...] = ()

    def to_mapping(self) -> dict[str, Any]:
        return {
            "argv": list(self.argv),
            "cwd": self.cwd,
            "materials": list(self.materials),
        }


@dataclass(frozen=True, slots=True)
class FinalAssuranceMaterial:
    """Historical repository-relative final-custody request.

    The current stable-result profile refuses new declarations because Zeroshot's
    delivery receipt, not a Broodling worktree read, is the accepted result.
    Keeping the type preserves exact historical Contract meaning.
    """

    path: str
    final_candidate: bool = True
    comparison_base: bool = False

    def to_mapping(self) -> dict[str, Any]:
        return {
            "path": self.path,
            "finalCandidate": self.final_candidate,
            "comparisonBase": self.comparison_base,
        }


@dataclass(frozen=True, slots=True)
class Criterion:
    """A required outcome with optional validation guidance for Zeroshot."""

    criterion_id: str
    statement: str
    evidence_population: EvidencePopulation | None = None
    validation_seam: str = ""
    validation_action: str = ""
    falsifying_observation: str = ""
    evidence_effect_dependencies: tuple[str, ...] = ()
    mechanical_evidence: MechanicalEvidence | None = None

    def to_mapping(self) -> dict[str, Any]:
        result = {
            "criterionId": self.criterion_id,
            "statement": self.statement,
            "validationSeam": self.validation_seam,
            "validationAction": self.validation_action,
            "falsifyingObservation": self.falsifying_observation,
            "evidenceEffectDependencies": list(self.evidence_effect_dependencies),
        }
        if self.evidence_population is not None:
            result["evidencePopulation"] = self.evidence_population.to_mapping()
        # Absence preserves the exact bytes and meaning of historical revisions.
        if self.mechanical_evidence is not None:
            result["mechanicalEvidence"] = self.mechanical_evidence.to_mapping()
        return result


@dataclass(frozen=True, slots=True)
class Obligation:
    """A stated obligation and its kind. Kinds outside the V1 profile are kept."""

    obligation_id: str
    statement: str
    kind: str

    def to_mapping(self) -> dict[str, Any]:
        return {
            "obligationId": self.obligation_id,
            "statement": self.statement,
            "kind": self.kind,
        }


@dataclass(frozen=True, slots=True)
class RequiredEffect:
    """An authoritative external effect the Contract requires.

    A pull-request effect may name its exact target branch. Other effects remain
    representable so admission can preserve and refuse them without weakening the
    Contract.
    """

    effect_id: str
    statement: str
    kind: str
    target_branch: str = ""

    def to_mapping(self) -> dict[str, Any]:
        result = {
            "effectId": self.effect_id,
            "statement": self.statement,
            "kind": self.kind,
        }
        # Omission preserves the exact canonical form of historical Contracts.
        if self.target_branch:
            result["targetBranch"] = self.target_branch
        return result


@dataclass(frozen=True, slots=True)
class Prerequisite:
    """A prerequisite the Contract needs inside the qualified V1 profile."""

    prerequisite_id: str
    statement: str
    satisfied_within_profile: bool

    def to_mapping(self) -> dict[str, Any]:
        return {
            "prerequisiteId": self.prerequisite_id,
            "statement": self.statement,
            "satisfiedWithinProfile": self.satisfied_within_profile,
        }


@dataclass(frozen=True, slots=True)
class Contract:
    """A complete Contract body, ready to be recorded as an immutable revision."""

    work_unit_id: str
    source_attribution: tuple[SourceAttribution, ...]
    criteria: tuple[Criterion, ...]
    obligations: tuple[Obligation, ...] = ()
    prerequisites: tuple[Prerequisite, ...] = ()
    required_effects: tuple[RequiredEffect, ...] = ()
    host_assumptions: tuple[str, ...] = ()
    constructed_by: str = "broodling_policy"
    notes: str = ""
    final_assurance_materials: tuple[FinalAssuranceMaterial, ...] | None = None

    def to_mapping(self) -> dict[str, Any]:
        result = {
            "contractFormat": CONTRACT_FORMAT,
            "workUnitId": self.work_unit_id,
            "constructedBy": self.constructed_by,
            "sourceAttribution": [
                item.to_mapping() for item in self.source_attribution
            ],
            "criteria": [item.to_mapping() for item in self.criteria],
            "obligations": [item.to_mapping() for item in self.obligations],
            "prerequisites": [item.to_mapping() for item in self.prerequisites],
            "requiredEffects": [item.to_mapping() for item in self.required_effects],
            "hostAssumptions": list(self.host_assumptions),
            "notes": self.notes,
        }
        # New selection is new revision meaning, never an amendment of old bytes.
        if self.final_assurance_materials is not None:
            result["finalAssuranceMaterials"] = [
                item.to_mapping() for item in self.final_assurance_materials
            ]
        return result

    def canonical_bytes(self) -> bytes:
        """Deterministic serialization; the durable meaning of the revision."""

        return json.dumps(
            self.to_mapping(),
            sort_keys=True,
            separators=(",", ":"),
            ensure_ascii=False,
        ).encode("utf-8")

    @property
    def contract_sha256(self) -> str:
        return digest("broodling.contract.v1", self.canonical_bytes().decode("utf-8"))

    @property
    def contract_revision_id(self) -> str:
        """Durable revision id. Identical meaning re-resolves the same revision."""

        return "cr-" + digest(
            "broodling.contract-revision.v1", self.work_unit_id, self.contract_sha256
        )


def contract_from_mapping(mapping: dict[str, Any]) -> Contract:
    """Rebuild a Contract from its canonical mapping.

    Used to read a stored revision back without a second source of truth for the
    field names.
    """

    if mapping.get("contractFormat") != CONTRACT_FORMAT:
        raise ValueError(
            f"unsupported contract format {mapping.get('contractFormat')!r}"
        )
    return Contract(
        work_unit_id=mapping["workUnitId"],
        constructed_by=mapping["constructedBy"],
        source_attribution=tuple(
            SourceAttribution(item["sourceId"], item["contentSha256"])
            for item in mapping["sourceAttribution"]
        ),
        criteria=tuple(
            Criterion(
                criterion_id=item["criterionId"],
                statement=item["statement"],
                evidence_population=(
                    EvidencePopulation(
                        kind=item["evidencePopulation"]["kind"],
                        members=tuple(item["evidencePopulation"]["members"]),
                        surface=item["evidencePopulation"]["surface"],
                    )
                    if "evidencePopulation" in item
                    else None
                ),
                validation_seam=item["validationSeam"],
                validation_action=item["validationAction"],
                falsifying_observation=item["falsifyingObservation"],
                evidence_effect_dependencies=tuple(item["evidenceEffectDependencies"]),
                mechanical_evidence=(
                    _mechanical_from_mapping(item["mechanicalEvidence"])
                    if "mechanicalEvidence" in item
                    else None
                ),
            )
            for item in mapping["criteria"]
        ),
        obligations=tuple(
            Obligation(item["obligationId"], item["statement"], item["kind"])
            for item in mapping["obligations"]
        ),
        prerequisites=tuple(
            Prerequisite(
                item["prerequisiteId"],
                item["statement"],
                item["satisfiedWithinProfile"],
            )
            for item in mapping["prerequisites"]
        ),
        required_effects=tuple(
            RequiredEffect(
                item["effectId"],
                item["statement"],
                item["kind"],
                item.get("targetBranch", ""),
            )
            for item in mapping["requiredEffects"]
        ),
        host_assumptions=tuple(mapping["hostAssumptions"]),
        notes=mapping["notes"],
        final_assurance_materials=(
            _final_materials_from_mapping(mapping["finalAssuranceMaterials"])
            if "finalAssuranceMaterials" in mapping
            else None
        ),
    )


def _final_materials_from_mapping(value: Any) -> tuple[FinalAssuranceMaterial, ...]:
    if not isinstance(value, list):
        raise ValueError("finalAssuranceMaterials must be an array")  # noqa: TRY004
    result = []
    for item in value:
        if not isinstance(item, dict) or set(item) != {
            "path",
            "finalCandidate",
            "comparisonBase",
        }:
            raise ValueError(
                "final assurance material requires exactly path, finalCandidate "
                "and comparisonBase"
            )
        result.append(
            FinalAssuranceMaterial(
                item["path"], item["finalCandidate"], item["comparisonBase"]
            )
        )
    return tuple(result)


def _mechanical_from_mapping(mapping: Any) -> MechanicalEvidence:
    """Refuse malformed structured meaning instead of dropping unknown fields."""

    if not isinstance(mapping, dict) or set(mapping) != {"argv", "cwd", "materials"}:
        raise ValueError("mechanicalEvidence requires exactly argv, cwd and materials")
    if not isinstance(mapping["argv"], list) or not isinstance(
        mapping["materials"], list
    ):
        # Invalid serialized Contract values use one fail-closed error surface.
        raise ValueError("mechanicalEvidence argv and materials must be arrays")  # noqa: TRY004
    return MechanicalEvidence(
        tuple(mapping["argv"]), mapping["cwd"], tuple(mapping["materials"])
    )

File: src/Broodling/Contract.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Broodling;

public sealed record SourceAttribution(string SourceId, string ContentSha256);
public sealed record Obligation(string ObligationId, string Statement, string Kind);
public sealed record RequiredEffect(string EffectId, string Statement, string Kind, string TargetBranch = "");
public sealed record Prerequisite(string PrerequisiteId, string Statement, bool SatisfiedWithinProfile);
public sealed record FinalAssuranceMaterial(string Path, bool FinalCandidate = true, bool ComparisonBase = false);

public sealed class EvidencePopulation(string kind, IReadOnlyList<string>? members = null, string surface = "")
{
    public string Kind { get; } = kind;
    public IReadOnlyList<string> Members { get; } = ContractData.Copy(members ?? [], nameof(members));
    public string Surface { get; } = surface;
}

public sealed class MechanicalEvidence(IReadOnlyList<string> argv, string cwd = ".", IReadOnlyList<string>? materials = null)
{
    public IReadOnlyList<string> Argv { get; } = ContractData.Copy(argv, nameof(argv));
    public string Cwd { get; } = cwd;
    public IReadOnlyList<string> Materials { get; } = ContractData.Copy(materials ?? [], nameof(materials));
}

public sealed class Criterion(string criterionId, string statement, EvidencePopulation? evidencePopulation = null,
    string validationSeam = "", string validationAction = "", string falsifyingObservation = "",
    IReadOnlyList<string>? evidenceEffectDependencies = null, MechanicalEvidence? mechanicalEvidence = null)
{
    public string CriterionId { get; } = criterionId;
    public string Statement { get; } = statement;
    public EvidencePopulation? EvidencePopulation { get; } = evidencePopulation;
    public string ValidationSeam { get; } = validationSeam;
    public string ValidationAction { get; } = validationAction;
    public string FalsifyingObservation { get; } = falsifyingObservation;
    public IReadOnlyList<string> EvidenceEffectDependencies { get; } = ContractData.Copy(evidenceEffectDependencies ?? [], nameof(evidenceEffectDependencies));
    public MechanicalEvidence? MechanicalEvidence { get; } = mechanicalEvidence;
}

/// <summary>Complete immutable meaning. Unsupported requirements remain representable for refusal.</summary>
public sealed class Contract
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow,
        RespectRequiredConstructorParameters = true
    };

    public Contract(string workUnitId, IReadOnlyList<SourceAttribution> sourceAttribution, IReadOnlyList<Criterion> criteria,
        IReadOnlyList<Obligation>? obligations = null, IReadOnlyList<Prerequisite>? prerequisites = null,
        IReadOnlyList<RequiredEffect>? requiredEffects = null, IReadOnlyList<string>? hostAssumptions = null,
        string constructedBy = "broodling_policy", string notes = "",
        IReadOnlyList<FinalAssuranceMaterial>? finalAssuranceMaterials = null)
    {
        WorkUnitId = workUnitId;
        // Attribution is a set of exact pins, not precedence. Keep duplicates so
        // validation can reject them instead of silently changing the proposal.
        SourceAttribution = Array.AsReadOnly(ContractData.Copy(sourceAttribution, nameof(sourceAttribution))
            .OrderBy(pin => pin.SourceId, StringComparer.Ordinal)
            .ThenBy(pin => pin.ContentSha256, StringComparer.Ordinal).ToArray());
        Criteria = ContractData.Copy(criteria, nameof(criteria));
        Obligations = ContractData.Copy(obligations ?? [], nameof(obligations));
        Prerequisites = ContractData.Copy(prerequisites ?? [], nameof(prerequisites));
        RequiredEffects = ContractData.Copy(requiredEffects ?? [], nameof(requiredEffects));
        HostAssumptions = ContractData.Copy(hostAssumptions ?? [], nameof(hostAssumptions));
        ConstructedBy = constructedBy;
        Notes = notes;
        FinalAssuranceMaterials = finalAssuranceMaterials is null ? null : ContractData.Copy(finalAssuranceMaterials, nameof(finalAssuranceMaterials));
    }

    public string WorkUnitId { get; }
    public IReadOnlyList<SourceAttribution> SourceAttribution { get; }
    public IReadOnlyList<Criterion> Criteria { get; }
    public IReadOnlyList<Obligation> Obligations { get; }
    public IReadOnlyList<Prerequisite> Prerequisites { get; }
    public IReadOnlyList<RequiredEffect> RequiredEffects { get; }
    public IReadOnlyList<string> HostAssumptions { get; }
    public string ConstructedBy { get; }
    public string Notes { get; }
    public IReadOnlyList<FinalAssuranceMaterial>? FinalAssuranceMaterials { get; }

    [JsonIgnore]
    public string ContractSha256 => Digests.Bytes(CanonicalBytes());
    [JsonIgnore]
    public string ContractRevisionId => "cr-" + ContractSha256;

    public byte[] CanonicalBytes()
    {
        Validate();
        return JsonSerializer.SerializeToUtf8Bytes(this, JsonOptions);
    }

    public static Contract FromCanonicalBytes(byte[] bytes)
    {
        try
        {
            var contract = JsonSerializer.Deserialize<Contract>(bytes, JsonOptions)
                ?? throw new InvalidContractProposal("Contract material must be an object.");
            // Stored material is our canonical format, not permissive model JSON.
            // Never silently discard unknown, omitted, or malformed meaning.
            if (!contract.CanonicalBytes().AsSpan().SequenceEqual(bytes))
                throw new InvalidContractProposal("Stored Contract material is not canonical.");
            return contract;
        }
        catch (JsonException)
        {
            throw new InvalidContractProposal("Contract material is malformed or has unrecognized fields.");
        }
    }

    public void Validate()
    {
        ContractData.Text(WorkUnitId, "Work Unit", required: true);
        ContractData.Text(Notes, "notes");
        if (ConstructedBy is not ("broodling_policy" or "caller" or "model_extraction"))
            throw new InvalidContractProposal("Unrecognized Contract producer.");
        var sources = new HashSet<string>(StringComparer.Ordinal);
        foreach (var pin in SourceAttribution)
        {
            ContractData.Text(pin.SourceId, "source identity", required: true);
            ContractData.Text(pin.ContentSha256, "source digest", required: true);
            if (!sources.Add(pin.SourceId))
                throw new SourceAttributionError("A source must be attributed once at its exact digest.");
        }
        ValidateStatements(Criteria.Select(item => (item.CriterionId, item.Statement)), "criterion");
        ValidateStatements(Obligations.Select(item => (item.ObligationId, item.Statement)), "obligation");
        ValidateStatements(Prerequisites.Select(item => (item.PrerequisiteId, item.Statement)), "prerequisite");
        ValidateStatements(RequiredEffects.Select(item => (item.EffectId, item.Statement)), "effect");
        foreach (var criterion in Criteria)
        {
            ContractData.Text(criterion.ValidationSeam, "validation seam");
            ContractData.Text(criterion.ValidationAction, "validation action");
            ContractData.Text(criterion.FalsifyingObservation, "falsifying observation");
            foreach (var dependency in criterion.EvidenceEffectDependencies)
                ContractData.Text(dependency, "evidence effect dependency");
            if (criterion.EvidencePopulation is { } population)
            {
                ContractData.Text(population.Kind, "evidence population kind");
                ContractData.Text(population.Surface, "evidence population surface");
                foreach (var member in population.Members)
                    ContractData.Text(member, "evidence population member");
            }
            if (criterion.MechanicalEvidence is { } mechanical)
            {
                ContractData.Text(mechanical.Cwd, "mechanical evidence directory");
                foreach (var argument in mechanical.Argv)
                    ContractData.Text(argument, "mechanical evidence argument");
                foreach (var material in mechanical.Materials)
                    ContractData.Text(material, "mechanical evidence material");
            }
        }
        foreach (var obligation in Obligations)
            ContractData.Text(obligation.Kind, "obligation kind");
        foreach (var effect in RequiredEffects)
        {
            ContractData.Text(effect.Kind, "effect kind");
            ContractData.Text(effect.TargetBranch, "target branch");
        }
        foreach (var assumption in HostAssumptions)
            ContractData.Text(assumption, "host assumption");
        foreach (var material in FinalAssuranceMaterials ?? [])
            ContractData.Text(material.Path, "selected material path");
    }

    private static void ValidateStatements(IEnumerable<(string Identity, string Statement)> items, string kind)
    {
        var identities = new HashSet<string>(StringComparer.Ordinal);
        foreach (var (identity, statement) in items)
        {
            ContractData.Text(identity, kind + " identity", required: true);
            ContractData.Text(statement, kind + " statement", required: true);
            if (!identities.Add(identity))
                throw new InvalidContractProposal($"{kind} identities must be unique.");
        }
    }
}

internal static class ContractData
{
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);

    internal static IReadOnlyList<T> Copy<T>(IReadOnlyList<T> values, string name) where T : class
    {
        if (values is null)
            throw new InvalidContractProposal($"{name} must be a collection.");
        var copy = values.ToArray();
        if (copy.Any(value => value is null))
            throw new InvalidContractProposal($"{name} cannot contain null elements.");
        return Array.AsReadOnly(copy);
    }

    internal static void Text(string value, string name, bool required = false)
    {
        if (value is null || (required && string.IsNullOrWhiteSpace(value)))
            throw new InvalidContractProposal($"{name} must be {(required ? "nonempty text" : "text")}.");
        try { StrictUtf8.GetByteCount(value); }
        catch (EncoderFallbackException)
        {
            throw new InvalidContractProposal($"{name} contains invalid Unicode.");
        }
    }
}
